kedro/io/data_catalog.py

```diff
@@ -257,36 +266,105 @@ class to be loaded is specified with the key ``type`` and their
             >>> catalog.save("boats", df)
         """
         data_sets = {}
+        dataset_patterns = {}
         catalog = copy.deepcopy(catalog) or {}
         credentials = copy.deepcopy(credentials) or {}
         save_version = save_version or generate_timestamp()
         load_versions = copy.deepcopy(load_versions) or {}
+        layers: dict[str, set[str]] = defaultdict(set)
 
-        missing_keys = load_versions.keys() - catalog.keys()
+        for ds_name, ds_config in catalog.items():
+            ds_config = _resolve_credentials(ds_config, credentials)
+            if cls._is_pattern(ds_name):
+                # Add each factory to the dataset_patterns dict.
+                dataset_patterns[ds_name] = ds_config
+
+            else:
+                ds_layer = ds_config.pop("layer", None)
+                if ds_layer is not None:
+                    layers[ds_layer].add(ds_name)
+                data_sets[ds_name] = AbstractDataSet.from_config(
+                    ds_name, ds_config, load_versions.get(ds_name), save_version
+                )
+        dataset_layers = layers or None
+        sorted_patterns = cls._sort_patterns(dataset_patterns)
+        missing_keys = [
+            key
+            for key in load_versions.keys()
+            if not (cls._match_pattern(sorted_patterns, key) or key in catalog)
+        ]
         if missing_keys:
             raise DatasetNotFoundError(
                 f"'load_versions' keys [{', '.join(sorted(missing_keys))}] "
                 f"are not found in the catalog."
             )
 
-        layers: dict[str, set[str]] = defaultdict(set)
-        for ds_name, ds_config in catalog.items():
-            ds_layer = ds_config.pop("layer", None)
-            if ds_layer is not None:
-                layers[ds_layer].add(ds_name)
+        return cls(
+            data_sets=data_sets,
+            layers=dataset_layers,
+            dataset_patterns=sorted_patterns,
+            load_versions=load_versions,
+            save_version=save_version,
+        )
 
-            ds_config = _resolve_credentials(ds_config, credentials)
-            data_sets[ds_name] = AbstractDataSet.from_config(
-                ds_name, ds_config, load_versions.get(ds_name), save_version
-            )
+    @staticmethod
```

Comment:
What's the reason behind some of these new methods being static methods and some class methods? None of these actually return a class instance.

Reply:
`_is_pattern`, `_match_pattern` & `_sort_patterns` (also `_specificity` which is used by `_sort_patterns` need to be either class methods or static methods because they're used in `from_config` which is a class method. Class/static methods do not necessarily need to return class instances, they just can be used without having an instance of the class.

Reply:
Ah sorry I don't think my question was very clear. I understand why these aren't regular instance methods, but my question is why `_is_pattern` and `_match_pattern` are static methods and why `_sort_patterns` and `_specificity` are class methods? Why aren't they all the same type?

Reply:
`_specificity`, `_match_pattern` and `_is_pattern` now are all static methods but `_sort_patterns()` still needs to be a class method because it uses `_specificity` inside it which is only accessible if we pass `cls` as an argument.

Reply:
Okay cool, thanks for clarifying 🙂